signalingserver.go

```diff
@@ -14,23 +14,161 @@
+	}
+	connId := dec.ConnId
+	dev, err := s.instanceManager.DeviceFromId(dec.DevId)
+	if err != nil {
+		return SServerResponse{StatusCode: http.StatusInternalServerError}, err
+	}
+	query := fmt.Sprintf("start=%d", start)
+	if count > 0 {
+		query = fmt.Sprintf("%s&count=%d", query, count)
+	}
+	var resErr ErrorMsg
+	var reply interface{}
+	status, err := GETRequest(hostURL(dev.Addr, "/polled_connections/"+connId+"/messages", query), &reply, &resErr)
+	return SServerResponse{reply, status}, err
+}
+
+func hostURL(addr string, path string, query string) string {
+	url := "http://" + addr + ":1080" + path
+	if query != "" {
+		url += "?" + query
+	}
+	return url
+}
+
+// Returns the http response's status code or an error.
+// If the status code indicates success (in the 2xx range) the response will be
+// in resObj, otherwise resErr will contain the error message.
+func POSTRequest(url string, msg interface{}, resObj interface{}, resErr *ErrorMsg) (int, error) {
+	jsonBody, err := json.Marshal(msg)
+	if err != nil {
+		log.Println("Error parsing json: ", err)
+		return http.StatusInternalServerError, err
+	}
+	reqBody := bytes.NewBuffer(jsonBody)
+	res, err := http.Post(url, "application/json", reqBody)
+	if err != nil {
+		log.Println("Error connecting to device host: ", err)
+		return http.StatusInternalServerError, err
+	}
+	defer res.Body.Close()
+	return parseReply(res, resObj, resErr)
+}
+
+func GETRequest(url string, resObj interface{}, resErr *ErrorMsg) (int, error) {
+	res, err := http.Get(url)
+	if err != nil {
+		log.Println("Error connecting to device host: ", err)
+		return http.StatusInternalServerError, err
+	}
+	defer res.Body.Close()
+	return parseReply(res, resObj, resErr)
+}
+
+func parseReply(res *http.Response, resObj interface{}, resErr *ErrorMsg) (int, error) {
+	var err error
+	dec := json.NewDecoder(res.Body)
+	if res.StatusCode < 200 || res.StatusCode > 299 {
+		err = dec.Decode(resErr)
+	} else {
+		err = dec.Decode(resObj)
+	}
+	if err != nil {
+		log.Println("Failed to parse response json: ", err)
+		return -1, err
+	}
+	return res.StatusCode, nil
+}
+
+const SEPARATOR string = ":"
+
+func encodeConnId(connId string, deviceId string) string {
+	// Both the device id and the connection id could have any characters in it,
+	// the connection id is base64 enocoded to ensure it doesn't have the
+	// separator ('/').
+	return deviceId + SEPARATOR + base64.StdEncoding.EncodeToString([]byte(connId))
 }
 
-func (s *ForwardingSignalingServer) Forward(id string, msg ForwardMsg) error {
-	return nil
+type IdDecodeResult struct {
+	ConnId string
+	DevId  string
 }
 
-func (s *ForwardingSignalingServer) Messages(id string, start int, count int) ([]interface{}, error) {
-	return make([]interface{}, 0), nil
+func decodeConnId(connId string) (IdDecodeResult, error) {
+	idx := strings.LastIndex(connId, SEPARATOR)
+	if idx < 0 {
+		return IdDecodeResult{}, errors.New("Malformed connection id")
```

Comment:
let's mention that a separator is missing and add the connId value.

Reply:
Done